core/build_playerglobal/src/lib.rs

```diff
@@ -62,3 +70,186 @@ pub fn build_playerglobal(
+            // For example, the freestanding function "flash.utils.getDefinitionByName"
+            // has a namespace of "flash.utils", which turns into the path
+            // "flash::utils"
+            path += &flash_to_rust_path(resolve_multiname_ns(&abc, trait_name));
+            path += "::";
+        }
+
+        // Append the trait name - this corresponds to the actual method
+        // name (e.g. `getDefinitionByName`)
+        path += method_prefix;
+        path += &flash_to_rust_path(resolve_multiname_name(&abc, trait_name));
+
+        // Now that we've built up the path, convert it into a `TokenStream`.
+        // This gives us something like
+        // `crate::avm2::globals::flash::system::Security::allowDomain`
+        //
+        // The resulting `TokenStream` is suitable for usage with `quote!` to
+        // generate a reference to the function pointer that should exist
+        // at that path in Rust code.
+        let path_tokens = TokenStream::from_str(&path).unwrap();
+        rust_paths[method_id.0 as usize] = quote! { Some(#path_tokens) };
+    };
+
+    // We support three kinds of native methods:
+    // instance methods, class methods, and freestanding functions.
+    // We're going to insert them into an `FnvHashMap`, so the order
+    // that we visit them in doesn't matter.
+    for (i, instance) in abc.instances.iter().enumerate() {
+        // Look for native instance methods
+        for trait_ in &instance.traits {
+            check_trait(trait_, Some(instance.name));
+        }
+        // Look for native class methods (in the corresponding
+        // `Class` definition)
+        for trait_ in &abc.classes[i].traits {
+            check_trait(trait_, Some(instance.name));
+        }
+    }
+
+    // Look for freestanding methods
+    for script in &abc.scripts {
+        for trait_ in &script.traits {
+            check_trait(trait_, None);
+        }
+    }
+
+    // Finally, generate the actual code. This is a Rust array -
+    // the entry at index `i` is a Rust function pointer for the native
+    // method with id `i`. Not all methods in playerglobal will be native
+    // methods, so we store `None` in the entries corresponding to non-native
+    // functions. We expect the majority of the methods in playerglobal to be
+    // native, so this should only waste a small amount of memory.
+    //
+    // If a function pointer doesn't exist at the expected path,
+    // then Ruffle compilation will fail
+    // with an error message that mentions the non-existent path.
+    //
+    // When we initially load a method from an ABC file, we check if it's from our playerglobal,
+    // and if its ID exists in this table.
+    // If so, we replace it with a `NativeMethod` constructed
+    // from the function pointer we looked up in the table.
+
+    let make_native_table = quote! {
+        const NATIVE_TABLE: &[Option<crate::avm2::method::NativeMethodImpl>] = &[
+            #(#rust_paths),*,
```

Comment:
nit: I think this can be `#(#rust_paths,)*` (i.e. with a single comma inside the parenthesis).